### PR TITLE
Use UUIDs for matching

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This partly sets some of the groundwork for more immutable segments in future, but really I think this solves the test fails I'm getting on #3658.

Replace `id()` matching with explicit `uuid` so that we can control where to match and where not.

### CODE REVIEW

File: src/sqlfluff/core/parser/segments/base.py [hunk 1]
```diff
@@ -237,6 +238,7 @@ def __init__(
         segments,
         pos_marker: Optional[PositionMarker] = None,
         name: Optional[str] = None,
+        uuid: Optional[UUID] = None,
```
Comment (barrywhart) [inline src/sqlfluff/core/parser/segments/base.py:241]: I'm curious: Under what circumstances would a caller provide this parameter?

File: src/sqlfluff/core/parser/segments/raw.py [hunk 2]
```diff
@@ -56,6 +58,8 @@ def __init__(
         self._is_expandable = None
         # Keep track of any source fixes
         self._source_fixes = source_fixes
+        # UUID for matching
```
Comment (barrywhart) [inline src/sqlfluff/core/parser/segments/raw.py:61]: Nitpick: Instead of this, should it call the class (`BaseSegment`) constructor instead?

**Review by barrywhart — APPROVED** [review]
Looks good. A couple of questions/suggestions.

File: src/sqlfluff/core/parser/segments/base.py (hunk 1 above)
Comment (alanmcruickshank) [reply src/sqlfluff/core/parser/segments/base.py:241]: It's effectively called during a `.copy()` operation - in theory you'd call it if you were ever making a new segment which you wanted to represent (or at least act as an anchor for) an existing segment. Going deeper into immutability might require that a little more.

File: src/sqlfluff/core/parser/segments/raw.py (hunk 2 above)
Comment (alanmcruickshank) [reply src/sqlfluff/core/parser/segments/raw.py:61]: Yeah I thought about that. On balance I thought so much is different about the two constructors that inheriting seemed the more complex route.

**Pull request merged**